File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/EventSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.joda.time.format.ISODateTimeFormat;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.Event;
import eu.pretix.libpretixsync.db.Migrations;
import eu.pretix.libpretixsync.utils.JSONUtils;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;

public class EventSyncAdapter extends BaseSingleObjectSyncAdapter<Event> {
    public EventSyncAdapter(BlockingEntityStore<Persistable> store, String eventSlug, String key, PretixApi api, String syncCycleId, SyncManager.ProgressFeedback feedback) {
        super(store, eventSlug, key, api, syncCycleId, feedback);
    }

    @Override
    public void updateObject(Event obj, JSONObject jsonobj) throws JSONException {
        obj.setSlug(jsonobj.getString("slug"));
        obj.setCurrency(jsonobj.getString("currency"));
        obj.setDate_from(ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_from")).toDate());
        if (!jsonobj.isNull("date_to")) {
            obj.setDate_to(ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_to")).toDate());
        }
        obj.setLive(jsonobj.getBoolean("live"));
        obj.setHas_subevents(jsonobj.getBoolean("has_subevents"));
        obj.setJson_data(jsonobj.toString());
    }

    Event getKnownObject() {
        List<Event> is = store.select(Event.class)
                .where(Event.SLUG.eq(key))
                .get().toList();
        if (is.size() == 0) {
            return null;
        } else if (is.size() == 1) {
            return is.get(0);
        } else {
            // What's going on here? Let's delete and re-fetch
            store.delete(is);
            return null;
        }
    }

    @Override
    protected String getUrl() {
        return api.organizerResourceUrl("events/" + key);
    }

    @Override
    String getResourceName() {
        return "events";
    }

    @Override
    Event newEmptyObject() {
        return new Event();
    }

    public void standaloneRefreshFromJSON(JSONObject data) throws JSONException {
        Event obj = store.select(Event.class)
                .where(Event.SLUG.eq(data.getString("slug")))
                .get().firstOr(newEmptyObject());
        JSONObject old = null;
        if (obj.getId() != null) {
            old = obj.getJSON();
        }
        // Store object
        data.put("__libpretixsync_dbversion", Migrations.CURRENT_VERSION);
        data.put("__libpretixsync_syncCycleId", syncCycleId);
        if (old == null) {
            updateObject(obj, data);
            store.insert(obj);
        } else {
            if (!JSONUtils.similar(data, old)) {
                updateObject(obj, data);
                store.update(obj);
            }
        }
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/EventSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.db.Migrations
import eu.pretix.libpretixsync.sqldelight.Event
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import eu.pretix.libpretixsync.utils.JSONUtils
import org.joda.time.format.ISODateTimeFormat
import org.json.JSONException
import org.json.JSONObject

class EventSyncAdapter(
    db: SyncDatabase,
    eventSlug: String,
    key: String,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback? = null,
) : SqBaseSingleObjectSyncAdapter<Event>(
    db = db,
    eventSlug = eventSlug,
    key = key,
    api = api,
    syncCycleId = syncCycleId,
    feedback = feedback,
) {

    override fun getKnownObject(): Event? {
        val known = db.eventQueries.selectBySlug(eventSlug).executeAsList()

        return if (known.isEmpty()) {
            null
        } else if (known.size == 1) {
            known[0]
        } else {
            // What's going on here? Let's delete and re-fetch
            db.eventQueries.deleteBySlug(eventSlug)
            null
        }
    }

    override fun getResourceName(): String = "events"

    override fun getUrl(): String = api.organizerResourceUrl("events/$key")

    override fun getJSON(obj: Event): JSONObject = JSONObject(obj.json_data!!)

    override fun insert(jsonobj: JSONObject) {
        val dateFrom =
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_from"))
                .toDate()

        val dateTo = if (!jsonobj.isNull("date_to")) {
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_to")).toDate()
        } else {
            null
        }

        db.eventQueries.insert(
            currency = jsonobj.getString("currency"),
            date_to = dateTo,
            date_from = dateFrom,
            has_subevents = jsonobj.getBoolean("has_subevents"),
            json_data = jsonobj.toString(),
            live = jsonobj.getBoolean("live"),
            slug = jsonobj.getString("slug"),
        )
    }

    override fun update(obj: Event, jsonobj: JSONObject) {
        val dateFrom =
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_from"))
                .toDate()

        val dateTo = if (!jsonobj.isNull("date_to")) {
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_to")).toDate()
        } else {
            null
        }

        db.eventQueries.updateFromJson(
            currency = jsonobj.getString("currency"),
            date_to = dateTo,
            date_from = dateFrom,
            has_subevents = jsonobj.getBoolean("has_subevents"),
            json_data = jsonobj.toString(),
            live = jsonobj.getBoolean("live"),
            slug = obj.slug,
        )
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.eventQueries.transaction(false, body)
    }

    @Throws(JSONException::class)
    fun standaloneRefreshFromJSON(data: JSONObject) {
        // Store object
        data.put("__libpretixsync_dbversion", Migrations.CURRENT_VERSION)
        data.put("__libpretixsync_syncCycleId", syncCycleId)
        val known = getKnownObject()
        if (known == null) {
            insert(data)
        } else {
            val old = JSONObject(known.json_data!!)
            if (!JSONUtils.similar(data, old)) {
                update(known, data)
            }
        }
    }
}
